src/fuzz_introspector/frontends/src/frontend_go.py

```diff
@@ -0,0 +1,610 @@
+# Copyright 2024 Fuzz Introspector Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+"""Fuzz Introspector Light frontend for Go"""
+
+from typing import Optional
+
+import os
+import pathlib
+import re
+
+import logging
+
+from tree_sitter import Language, Parser, Node
+import tree_sitter_go
+import yaml
+
+logger = logging.getLogger(name=__name__)
+LOG_FMT = '%(asctime)s.%(msecs)03d %(levelname)s %(module)s - %(funcName)s: %(message)s'
+
+
+class SourceCodeFile():
+    """Class for holding file-specific information."""
+
+    def __init__(self,
+                 source_file: str,
+                 source_content: Optional[bytes] = None):
+        logger.info('Processing %s' % source_file)
+
+        self.root = None
+        self.imports = []
+        self.source_file = source_file
+        self.tree_sitter_lang = Language(tree_sitter_go.language())
+        self.parser = Parser(self.tree_sitter_lang)
+
+        if source_content:
+            self.source_content = source_content
+        else:
+            with open(self.source_file, 'rb') as f:
+                self.source_content = f.read()
+
+        # List of function definitions in the source file.
+        self.functions = []
+        self.methods = []
+
+        # Initialization ruotines
+        self.load_tree()
+
+        # Load function/method declaration
+        self._set_function_declaration()
+        self._set_method_declaration()
+
+        # Parse import package
+        self._set_imports()
+
+    def load_tree(self):
+        """Load the the source code into a treesitter tree, and set
+        the root node."""
+        self.root = self.parser.parse(self.source_content).root_node
+
+    def _set_function_declaration(self):
+        """Internal helper for retrieving all functions."""
+        func_query_str = '( function_declaration ) @fd '
+        func_query = self.tree_sitter_lang.query(func_query_str)
+
+        function_res = func_query.captures(self.root)
+        for _, funcs in function_res.items():
+            for func in funcs:
+                self.functions.append(
+                    FunctionMethod(func, self.tree_sitter_lang, self, True))
+
+    def _set_method_declaration(self):
+        """Internal helper for retrieving all methods."""
+        func_query_str = '( method_declaration ) @fd '
```

Comment:
what is the difference between `method_declaration` and `function_declaration`?

Reply:
From go tree definition, it contains both method and function declaration. Which both of them needed to be handled in order to covers all of the functions and methods.
```
(
  (comment)* @doc
  .
  (function_declaration
    name: (identifier) @name) @definition.function
  (#strip! @doc "^//\\s*")
  (#set-adjacent! @doc @definition.function)
)

(
  (comment)* @doc
  .
  (method_declaration
    name: (field_identifier) @name) @definition.method
  (#strip! @doc "^//\\s*")
  (#set-adjacent! @doc @definition.method)
)
```